test/compressionworker.ts

```diff
@@ -0,0 +1,222 @@
+      let uploadMessage = message as IUploadMessage;
+      let payload = JSON.parse(uploadMessage.rawData);
+      assert.equal(payload.events.length, 1);
+      assert.equal(payload.events[0].type, MockEventName);
+      done();
+    }
+  });
+
+  it("validates that queued payload is sent on terminate", (done: DoneFn) => {
+    let worker = createTestWorker();
+    let mockEvent = getMockEvent();
+    let addEventMessage = createAddEventMessage(mockEvent);
+    let terminateMsg: ITimestampedWorkerMessage = {
+      type: WorkerMessageType.Terminate,
+      time: new Date().getTime()
+    };
+
+    assert.equal(workerMessages.length, 0);
+    messageHandlers.push(messageHandler);
+    worker.postMessage(JSON.stringify(addEventMessage));
+    worker.postMessage(JSON.stringify(terminateMsg));
+    scheduleTestFailureTimeout();
+
+    function messageHandler(message: IWorkerMessage) {
+      assert.equal(message.type, WorkerMessageType.Upload);
+      let uploadMessage = message as IUploadMessage;
+      let payload = JSON.parse(uploadMessage.rawData);
+      assert.equal(payload.events.length, 1);
+      assert.equal(payload.events[0].type, MockEventName);
+      done();
+    }
+  });
+
+  it("validates that payloads consisting of a single XHR error event are not uploaded", (done: DoneFn) => {
+    let worker = createTestWorker();
+    let mockXhrErrorEvent = getMockEvent(InstrumentationEventName);
+    let mockXhrErrorEventState = {
+      type: Instrumentation.XhrError
+    } as IXhrErrorEventState;
+    mockXhrErrorEvent.state = mockXhrErrorEventState;
+    let addEventMessage: IAddEventMessage = {
+      type: WorkerMessageType.AddEvent,
+      time: new Date().getTime(),
+      event: mockXhrErrorEvent
+    };
+    let forceUploadMessage = createForceUploadMessage();
+
+    messageHandlers.push(messageHandler);
+    worker.postMessage(JSON.stringify(addEventMessage));
+    worker.postMessage(JSON.stringify(forceUploadMessage));
+    scheduleTestSuccessTimeout(done);
+
+    function messageHandler(message: IWorkerMessage) {
+      // This handler should NOT be invoked, so intentional assertion failure in this path
+      assert.equal(true, false);
+      done();
+    }
+  });
+
+  function onWorkerMessage(evt: MessageEvent) {
+    let message = JSON.parse(evt.data) as IWorkerMessage;
+    workerMessages.push(message);
+    for (let i = 0; i < messageHandlers.length; i++) {
+      messageHandlers[i](message);
+    }
+  }
+
+  function createTestWorker() {
+    let worker = createCompressionWorker(getMockEnvelope(), onWorkerMessage);
+    (worker as any).isTestWorker = true;
+    return worker;
+  }
+
+  function endTestByTimeout(endWithSuccess: boolean, done?: DoneFn) {
+    assert.equal(true, !!endWithSuccess);
+    if (done) {
+      done();
+    }
+  }
+
+  function scheduleTestSuccessTimeout(done: DoneFn) {
+    jasmine.clock().uninstall();
+    testFailureTimeout = setTimeout(() => {
+      endTestByTimeout(true, done);
+    }, 1000);
```

Comment:
Seems brittle.

Reply:
Why? 1 second should be safe upper limit for communication between treads in a test environment. I moved this value to be a constant at the top of the file.

---
In reply to: [136876709](https://github.com/Microsoft/clarity-js/pull/28#discussion_r136876709) [](ancestors = 136876709)